README.md
#### ⚠️ This project is no longer maintained

# fetch-gists

[![Version](https://img.shields.io/npm/v/fetch-gists.svg?style=flat-square)](https://www.npmjs.com/package/fetch-gists)
[![Build Status](https://img.shields.io/github/actions/workflow/status/mike182uk/fetch-gists/ci.yml?branch=main&style=flat-square)](https://github.com/mike182uk/fetch-gists/actions/workflows/ci.yml?query=workflow%3ACI)
[![Coveralls](https://img.shields.io/coveralls/mike182uk/fetch-gists/main.svg?style=flat-square)](https://coveralls.io/r/mike182uk/fetch-gists)
[![npm](https://img.shields.io/npm/dm/fetch-gists.svg?style=flat-square)](https://www.npmjs.com/package/fetch-gists)
[![License](https://img.shields.io/github/license/mike182uk/fetch-gists.svg?style=flat-square)](https://www.npmjs.com/package/fetch-gists)

Fetch all of your gists from GitHub.

`fetch-gists` will handle all of the API calls needed to retrieve all of your gists from GitHub. This includes navigating the paginated results.

## Prerequisites

- GitHub account (duh!)
- GitHub [access token](https://github.com/blog/1509-personal-api-tokens) with the `gist` scope enabled

## Installation

```bash
npm install --save fetch-gists
```

## Usage

```js
const fetchGists = require('fetch-gists')

const accessToken = '<your-github-access-token>'

try {
  const gists = await fetchGists(accessToken)

  // gists successfully retrieved
} catch (err) {
  // something went wrong
}
```

`fetchGists` will return a promise. The promise will resolve once all gists for the account have been retrieved. Any errors that occur during the retrieval of the gists will cause the promise to reject.


CHANGELOG.md
# Changelog

# 5.0.0

- Drop support for Node.js `< 12.0.0`

# 4.0.0

- Drop support for Node.js `< 10.0.0`
- Fix GitHub [API authentication](https://developer.github.com/changes/2020-02-10-deprecating-auth-through-query-param/)

# 3.0.0

- Drop support for Node.js `< 8.0.0`

# 2.0.0

- Drop support for Node.js `< 4.0.0`
- Remove dependency on `bluebird`
- Error is returned if an access token is not provided

# 1.0.1

- Refactor internals to be more efficient

# 1.0.0

First release


test.js
const nock = require('nock')
const proxyquire = require('proxyquire')
const test = require('tape')

const ACCESS_TOKEN = 'TEST123'

test('retrieves all gists', async t => {
  nock.cleanAll()

  t.plan(1)

  const fetchGists = getFetchGists()
  const gists = [{ id: 1 }, { id: 2 }, { id: 3 }]

  mockGetGistsApiCall(1, gists, true)

  const result = await fetchGists(ACCESS_TOKEN)

  t.deepEqual(result, gists, 'gists retrieved')
})

test('follows pagination', async t => {
  nock.cleanAll()

  t.plan(1)

  const fetchGists = getFetchGists()
  const page1Gists = [{ id: 1 }, { id: 2 }, { id: 3 }]
  const page2Gists = [{ id: 4 }, { id: 5 }]
  const page3Gists = [{ id: 6 }]
  const allGists = [].concat(page1Gists, page2Gists, page3Gists)

  mockGetGistsApiCall(1, page1Gists, false)
  mockGetGistsApiCall(2, page2Gists, false)
  mockGetGistsApiCall(3, page3Gists, true)

  const result = await fetchGists(ACCESS_TOKEN)

  t.deepEqual(allGists, result, 'pagination links followed')
})

test('error when access token not provided', async t => {
  t.plan(1)

  const fetchGists = getFetchGists()

  try {
    await fetchGists()

    t.fail('expected fetchGists to throw')
  } catch ({ message }) {
    t.equal(message, 'You must supply an access token to retrieve your gists', 'error occured and contains correct message')
  }
})

test('error occurs on 401', async t => {
  nock.cleanAll()

  t.plan(2)

  const fetchGists = getFetchGists()
  const requestBody = {
    message: 'foo',
    documentation_url: 'bar'
  }

  mockGetGistsApiCall(1, requestBody, true, 401)

  try {
    await fetchGists(ACCESS_TOKEN)

    t.fail('expected fetchGists to throw')
  } catch ({ message }) {
    t.equal(message.includes(requestBody.message), true, 'error occured and contains message')
    t.equal(message.includes(requestBody.documentation_url), true, 'error occured and contains documentation link')
  }
})

test('error occurs on 403', async t => {
  nock.cleanAll()

  t.plan(2)

  const fetchGists = getFetchGists()
  const requestBody = {
    message: 'foo',
    documentation_url: 'bar'
  }

  mockGetGistsApiCall(1, requestBody, true, 403)

  try {
    await fetchGists(ACCESS_TOKEN)

    t.fail('expected fetchGists to throw')
  } catch ({ message }) {
    t.equal(message.includes(requestBody.message), true, 'error occured and contains message')
    t.equal(message.includes(requestBody.documentation_url), true, 'error occured and contains documentation link')
  }
})

test('error occurs when response code is not 200, 401 or 403', async t => {
  nock.cleanAll()

  t.plan(1)

  const fetchGists = getFetchGists()

  mockGetGistsApiCall(1, {}, true, 404)

  try {
    await fetchGists(ACCESS_TOKEN)

    t.fail('expected fetchGists to throw')
  } catch ({ message }) {
    t.equal(message.includes(404), true, 'error occured and contains status code')
  }
})

test('error occurs when request fails', async t => {
  nock.cleanAll()

  t.plan(1)

  mockGetGistsApiCall(1, [], true, 200)

  const errMsg = 'foo'
  const fetchGists = getFetchGists({
    got: () => Promise.reject(new Error(errMsg))
  })

  try {
    await fetchGists(ACCESS_TOKEN)

    t.fail('expected fetchGists to throw')
  } catch ({ message }) {
    t.equal(message, errMsg, 'error occured during request')
  }
})

function mockGetGistsApiCall (page, body, isLastPage, statusCode = 200) {
  const responseHeaders = {}

  if (!isLastPage) {
    responseHeaders.Link = `<link?page=${page + 1}> rel="next", <link?page=99> rel="last"`
  }

  const requestHeaders = {
    'User-Agent': 'fetch-gists',
    accept: 'application/vnd.github.v3+json',
    Authorization: `token ${ACCESS_TOKEN}`
  }

  nock('https://api.github.com', { reqheaders: requestHeaders })
    .get('/gists')
    .query({
      page,
      per_page: 100
    })
    .reply(statusCode, body, responseHeaders)
}

function getFetchGists (dependencies) {
  return proxyquire('./index.js', dependencies || {})
}


index.js
const got = require('got')

/**
 * Exports
 */

module.exports = fetchGists

/**
 * Fetch gists for the given access token
 *
 * @param   {string} accessToken
 * @returns {Promise.<Object[]>}
 */

async function fetchGists (accessToken) {
  if (!accessToken) {
    throw new Error('You must supply an access token to retrieve your gists')
  }

  let gists = []
  let page = 1
  let hasMore = true

  while (hasMore) {
    try {
      const { body, headers } = await got('https://api.github.com/gists', {
        headers: {
          'User-Agent': 'fetch-gists',
          accept: 'application/vnd.github.v3+json',
          Authorization: `token ${accessToken}`
        },
        searchParams: {
          page,
          per_page: 100
        },
        responseType: 'json'
      })

      gists = gists.concat(body)

      hasMore = headers.link
        ? headers.link.includes('rel="next"')
        : false

      page++
    } catch (err) {
      if (err.response) {
        const statusCode = err.response.statusCode
        const body = err.response.body

        if ([401, 403].includes(statusCode)) {
          throw new Error(`${body.message}. You can view the documentation at ${body.documentation_url}`)
        }

        if (statusCode !== 200) {
          throw new Error(`Expected 200 response code but got ${statusCode}`)
        }
      }

      throw err
    }
  }

  return gists
}
